### PR TITLE
Virtualize extension components

### LINKED ISSUE #3288: Virtualize Autocomplete and SelectList
Similar to DataGrid, the feature should be opt-in by a parameter.

### PR DESCRIPTION
**Pull request by stsrki**
Closes #3288

Virtualized:

- Autocomplete
- DropdownList
- ListView

SelectList cannot be virtualized because of https://github.com/dotnet/aspnetcore/issues/29087

### CODE REVIEW

**Comment by David-Moreira** [comment]
What a nice feature to have, you're on fire! :D

File: Source/Extensions/Blazorise.Components/Autocomplete.razor [hunk 1]
```diff
@@ -49,50 +50,47 @@
...
         {
-            for ( int i = 0; i < FilteredData.Count; ++i )
+            @if ( Virtualize )
             {
-                var item = FilteredData[i];
-                var isActiveItem = IsSuggestedActiveItem( item );
-                var isFocusedItem = i == ActiveItemIndex;
+                <Virtualize TItem="TItem" Context="item" Items="@FilteredData">
+                    @itemFragment(item)
+                </Virtualize>
```
Comment (David-Moreira) [inline Source/Extensions/Blazorise.Components/Autocomplete.razor:57]: **ReadData** support by implementing the appropriate handler?

File: Source/Extensions/Blazorise.Components/Autocomplete.razor.cs [hunk 2]
```diff
@@ -1415,5 +1415,10 @@ public List<string> SelectedTexts
     /// </summary>
     [Parameter] public bool AutoSelectFirstItem { get; set; }
 
+    /// <summary>
+    /// Gets or sets whether the autocomplete will use the Virtualize functionality.
+    /// </summary>
+    [Parameter] public bool Virtualize { get; set; }
+
```
Comment (David-Moreira) [inline Source/Extensions/Blazorise.Components/Autocomplete.razor.cs:1422]: Like in DataGrid, make available a **VirtualizeOptions** parameter across all components that use **Virtualzie**?

File: Source/Extensions/Blazorise.Components/DropdownList.razor [hunk 3]
```diff
@@ -1,18 +1,33 @@
...
         }
     </DropdownMenu>
 </Dropdown>
+@code {
+    protected RenderFragment<TItem> itemFragment => item => __builder =>
+    {
+        var text = TextField?.Invoke( item );
+        var value = ValueField != null ? ValueField.Invoke( item ) : default;
 
+        <DropdownItem Clicked="@HandleDropdownItemClicked" Value="@value">@text</DropdownItem>
```
Comment (David-Moreira) [inline Source/Extensions/Blazorise.Components/DropdownList.razor:31]: makes sense to have @key here?

File: Source/Extensions/Blazorise.Components/DropdownList.razor [hunk 4]
```diff
@@ -1,18 +1,33 @@
...
+﻿@using Microsoft.AspNetCore.Components.Web.Virtualization
+@typeparam TItem
 @typeparam TValue
 <Dropdown @ref="@dropdownRef" ElementId="@ElementId" Class="@Class" Style="@Style" RightAligned="@RightAligned" Disabled="@Disabled" Direction="@Direction" Attributes="@Attributes">
     <DropdownToggle @ref="@dropdownToggleRef" Color="@Color" TabIndex="@TabIndex">@ChildContent</DropdownToggle>
     <DropdownMenu MaxMenuHeight="@MaxMenuHeight">
         @if ( Data != null )
         {
-            foreach ( var item in Data )
+            @if ( Virtualize && Data is ICollection<TItem> collectionableData )
```
Comment (David-Moreira) [inline Source/Extensions/Blazorise.Components/DropdownList.razor:9]: Do we not have an util for this? Or we very specifically only want the ICollection derivative?

File: Source/Extensions/Blazorise.Components/ListView.razor [hunk 5]
```diff
@@ -1,20 +1,36 @@
...
-                <ListGroupItem @key="item" Name="@itemValue">@ItemTemplate( new( item, itemText, itemValue ) )</ListGroupItem>
-            }
-            else
+            <Virtualize TItem="TItem" Context="item" Items="@collectionableData">
+                @itemFragment(item)
+            </Virtualize>
+        }
+        else
+        {
+            @foreach ( var item in Data ?? Enumerable.Empty<TItem>() )
```
Comment (David-Moreira) [inline Source/Extensions/Blazorise.Components/ListView.razor:14]: I'm seeing this null check here, but I believe I saw some other similar component without the null check, maybe it was DropdownList? Should it be the same?

File: Source/Extensions/Blazorise.Components/ListView.razor [hunk 6]
```diff
@@ -1,20 +1,36 @@
...
+    {
+        var itemText = GetItemText( item );
+        var itemValue = GetItemValue( item );
+        @if ( ItemTemplate is not null )
+        {
+            <ListGroupItem @key="item" Name="@itemValue">@ItemTemplate( new( item, itemText, itemValue ) )</ListGroupItem>
+        }
+        else
+        {
+            <ListGroupItem @key="item" Name="@itemValue">@itemText</ListGroupItem>
```
Comment (David-Moreira) [inline Source/Extensions/Blazorise.Components/ListView.razor:33]: If we think / assume, itemvalue will always be unique, we could change the key to that, we did assume that in SelectList apparently.

File: Documentation/Blazorise.Docs/Pages/Docs/Extensions/Autocomplete/AutocompletePage.razor [hunk 7]
```diff
@@ -301,4 +301,7 @@
     <DocsAttributesItem Name="MaxEntryLength">
         Specifies the maximum number of characters allowed in the input element.
     </DocsAttributesItem>
+    <DocsAttributesItem Name="Virtualize" Type="bool" Default="false">
+        Gets or sets whether the autocomplete will use the <Code>Virtualize</Code> functionality.
```
Comment (David-Moreira) [inline Documentation/Blazorise.Docs/Pages/Docs/Extensions/Autocomplete/AutocompletePage.razor:305]: Capitalize the component if you think we should bother, same for the others.

**Comment by stsrki** [comment]
> What a nice feature to have, you're on fire! :D

image

**Review by David-Moreira — CHANGES_REQUESTED** [review]
Looking good, but we need more visibility on our features otherwise they can go unnoticed, make examples for these?

Missing release notes.

File: Source/Extensions/Blazorise.Components/DropdownList.razor (hunk 4 above)
Comment (stsrki) [reply Source/Extensions/Blazorise.Components/DropdownList.razor:9]: I wish, but `Virtualize` accepts `ICollection` type only.

File: Source/Extensions/Blazorise.Components/Autocomplete.razor (hunk 1 above)
Comment (stsrki) [reply Source/Extensions/Blazorise.Components/Autocomplete.razor:57]: You have more experience with this. Do you want to try it?

Comment (David-Moreira) [reply Source/Extensions/Blazorise.Components/Autocomplete.razor:57]: Tested with **ReadData** before starting to work on this and works already really good already.

But if there's huge ammounts of data, the ItemsProvider should be more performant as you'll only be querying that data specifically (similar to paging), but for rendering purposes it's already taking the advantage of the **Virtualization**.

Comment (David-Moreira) [reply Source/Extensions/Blazorise.Components/Autocomplete.razor]: Done, will you proceed with the rest of the PR? :)

Comment (stsrki) [reply Source/Extensions/Blazorise.Components/Autocomplete.razor]: I guess I will :)

File: Source/Extensions/Blazorise.Components/ListView.razor (hunk 6 above)
Comment (stsrki) [reply Source/Extensions/Blazorise.Components/ListView.razor:33]: The user should handle the uniqueness of data.

File: Source/Extensions/Blazorise.Components/Autocomplete.razor.cs (hunk 2 above)
Comment (stsrki) [reply Source/Extensions/Blazorise.Components/Autocomplete.razor.cs]: I guess we don't need this.

**Comment by stsrki** [comment]
@David-Moreira anything else?

**Comment by David-Moreira** [comment]
Maybe examples for the other components? Other than that nope.

**Review by David-Moreira — APPROVED** [review]

**Pull request merged**